src/ucp/core/ucp_worker.c

```diff
@@ -235,6 +283,42 @@ static void ucp_worker_set_stub_config(ucp_worker_h worker)
     config->sync_rndv_thresh  = SIZE_MAX;
 }
 
+static ucs_status_t ucs_wakeup_context_init(ucp_worker_wakeup_t *wakeup,
+                                            ucp_rsc_index_t num_tls)
+{
+    ucs_status_t status;
+
+    wakeup->iface_wakeups = ucs_calloc(num_tls, sizeof(*wakeup->iface_wakeups),
+                                       "ucp iface_wakeups");
+    if (wakeup->iface_wakeups == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (pipe(wakeup->wakeup_pipe) != 0) {
+        ucs_free(wakeup->iface_wakeups);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    status = ucs_sys_fcntl_modfl(wakeup->wakeup_pipe[0], O_NONBLOCK, 0);
```

Comment:
need to modify write side to non-blocking as well, so ucp_worker_signal would not block